tensorflow-ops/src/TensorFlow/Convolution.hs

```diff
@@ -0,0 +1,299 @@
+-- Copyright 2020 TensorFlow authors.
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module TensorFlow.Convolution
+    ( Padding(..)
+    , DataFormat(..)
+    , conv2D
+    , conv2D'
+    , conv2DBackpropFilter
+    , conv2DBackpropFilter'
+    , conv2DBackpropInput
+    , conv2DBackpropInput'
+    , conv3D
+    , conv3D'
+    , conv3DBackpropFilter
+    , conv3DBackpropFilter'
+    , conv3DBackpropFilterV2
+    , conv3DBackpropFilterV2'
+    , conv3DBackpropInput
+    , conv3DBackpropInput'
+    , conv3DBackpropInputV2
+    , conv3DBackpropInputV2'
+    , depthwiseConv2dNative
+    , depthwiseConv2dNative'
+    , depthwiseConv2dNativeBackpropFilter
+    , depthwiseConv2dNativeBackpropFilter'
+    , depthwiseConv2dNativeBackpropInput
+    , depthwiseConv2dNativeBackpropInput'
+    ) where
+
+import Data.Word(Word16)
+import Data.Int(Int32,Int64)
+import Data.ByteString(ByteString)
+import Lens.Family2 ((.~), (&))
+
+import qualified TensorFlow.BuildOp as TF
+import qualified TensorFlow.Core as TF
+import qualified TensorFlow.GenOps.Core as TF
+
+-- TODO: Support other convolution parameters such as stride.
+
+-- | Convolution padding.
+data Padding = PaddingValid -- ^ output_spatial_shape[i] = ceil(input_spatial_shape[i] / strides[i])
+             | PaddingSame  -- ^ output_spatial_shape[i] = ceil((input_spatial_shape[i] - (spatial_filter_shape[i]-1) * dilation_rate[i]) / strides[i])
```

Comment:
This pushed us over the github review line limit. Could you change to `-- |` haddock which should also be valid before the constructor name (I think).